File: choir/lib/core/Location.cc
#include <choir/macros.hh>
#include <choir/core.hh>
#include <llvm/ADT/StringRef.h>

using namespace choir;

bool Location::seekable(const Context& ctx) const {
    auto* f = ctx.file(file_id);
    if (not f) return false;
    return pos + len <= f->size() + 1 and is_valid();
}

/// Seek to a source location. The location must be valid.
auto Location::seek(const Context& ctx) const -> std::optional<LocInfo> {
    if (not seekable(ctx)) return std::nullopt;
    LocInfo info{};

    // Get the file that the location is in.
    const auto* f = ctx.file(file_id);

    // Seek back to the start of the line.
    const char* const data = f->data();
    info.line_start = data + pos;
    while (info.line_start > data and *info.line_start != '\n') info.line_start--;
    if (*info.line_start == '\n') info.line_start++;

    // Seek forward to the end of the line.
    const char* const end = data + f->size();
    info.line_end = data + pos;
    while (info.line_end < end and *info.line_end != '\n') info.line_end++;

    // Determine the line and column number.
    info.line = 1;
    info.col = 1;
    for (const char* d = data; d < data + pos; d++) {
        if (d < end and *d == '\n') {
            info.line++;
            info.col = 1;
        } else {
            info.col++;
        }
    }

    // Done!
    return info;
}

/// TODO: Lexer should create map that counts where in a file the lines start so
/// we can do binary search on that instead of iterating over the entire file.
auto Location::seek_line_column(const Context& ctx) const -> std::optional<LocInfoShort> {
    if (not seekable(ctx)) return std::nullopt;
    LocInfoShort info{};

    // Get the file that the location is in.
    const auto* f = ctx.file(file_id);

    // Seek back to the start of the line.
    const char* const data = f->data();
    const char* const end = data + f->size();

    // Determine the line and column number.
    info.line = 1;
    info.col = 1;
    for (const char* d = data; d < data + pos; d++) {
        if (d < end and *d == '\n') {
            info.line++;
            info.col = 1;
        } else {
            info.col++;
        }
    }

    // Done!
    return info;
}

auto Location::text(const Context& ctx) const -> llvm::StringRef {
    if (not seekable(ctx)) return "";
    auto* f = ctx.file(file_id);
    return llvm::StringRef{f->data(), size_t(f->size())}.substr(pos, len);
}

File: choir/bootstrap/Choir.Driver/Location.cs
using System.Security.Cryptography;

namespace Choir;

public struct Location(int offset, int length, int fileId)
{
    public static readonly Location Nowhere = new(0, 0, 0);

    public int Offset = offset;
    public int Length = length;
    public readonly int FileId = fileId;

    public readonly bool Seekable(ChoirContext context)
    {
        if (FileId <= 0) return false;
        var file = context.GetSourceFileById(FileId);
        if (file is null) return false;
        return Offset >= 0 && Offset + Length <= file.Text.Length;
    }

    public readonly LocationInfo? Seek(ChoirContext context)
    {
        if (!Seekable(context)) return null;
        var file = context.GetSourceFileById(FileId);
        if (file is null) return null;

        int lineStart = Offset;
        int lineEnd = Offset;

        string text = file.Text;

        // seek to start of line
        while (lineStart > 0 && text[lineStart] != '\n') lineStart--;
        if (text[lineStart] == '\n') lineStart++;

        // seek to end of line
        while (lineEnd < text.Length && text[lineEnd] != '\n') lineEnd++;

        int line = 1;
        int column = 1;

        for (int i = 0; i < Offset; i++)
        {
            if (text[i] == '\n')
            {
                line++;
                column = 1;
            }
            else column++;
        }

        return new(line, column, lineStart, lineEnd - lineStart, text.Substring(lineStart, lineEnd - lineStart));
    }

    public readonly LocationInfoShort? SeekLineColumn(ChoirContext context)
    {
        if (!Seekable(context)) return null;
        var file = context.GetSourceFileById(FileId)!;

        int line = 1;
        int column = 1;

        string text = file.Text;
        for (int i = 0; i < Offset; i++)
        {
            if (text[i] == '\n')
            {
                line++;
                column = 1;
            }
            else column++;
        }

        return new(line, column);
    }

    public readonly ReadOnlySpan<char> Span(ChoirContext context)
    {
        var file = context.GetSourceFileById(FileId);
        if (file is not null) return file.GetSpan(this);
        return "";
    }
}

public readonly struct LocationInfo(int line, int column, int lineStart, int lineLength, string lineText)
{
    public readonly int Line = line;
    public readonly int Column = column;
    public readonly int LineStart = lineStart;
    public readonly int LineLength = lineLength;
    public readonly string LineText = lineText;
}

public readonly struct LocationInfoShort(int line, int column)
{
    public readonly int Line = line;
    public readonly int Column = column;
}
